File: comms/native/services/networks/AStarTraceNode.js
// Copyright: IQGeo Limited 2010-2023

import { TraceNode } from 'myWorld-native-services';
import geomUtils from '../base/GeomUtils';
import { geometry } from 'myWorld-client';

class AStarTraceNode extends TraceNode {
    /**
     * Trace node for use with A* algorithm
     */

    /**
     * Comparison operator (used in heap operations)
     */
    // Subclassed to support ordering by distance to stop nodes (part of the A* algorithm implementation)
    cmp(other) {
        if (this.minPossibleDist !== undefined) {
            return this.minPossibleDist - other.minPossibleDist;
        }

        return this.dist - other.dist;
    }

    /**
     * Distance from self to nearest vertex on GEOMS, in m (0.0 if not known)
     *
     * GEOMS is a set of shapely geometries
     */
    minDistanceTo(geoms) {
        const self_coord = this._stop_coord();

        if (!self_coord) {
            return 0.0;
        }

        const p1 = geometry.point(self_coord);

        let min_dist = undefined;
        for (const geom of geoms) {
            const dist = p1.distanceTo(geom);

            if (min_dist == undefined || min_dist > dist) {
                min_dist = dist;
            }
        }

        return min_dist;
    }

    /**
     * Position on self's feature at which self ends
     */
    _stop_coord() {
        if (!this.parent)
            // ENH: Do better if root node
            return undefined;

        const geom = this.feature.geometry;
        if (geom.type == 'Point') {
            return geom.coordinates;
        }

        // Find position of stop point along feature (as proportion of total length)
        // Remember: dist may have been computed from a stored length value
        let pos;
        if (this.partial) {
            pos = (this.dist - this.parent.dist) / (this.full_dist - this.parent.dist);
        } else {
            pos = 1.0;
        }

        if (!this.isForward(geom)) {
            pos = 1.0 - pos;
        }

        return geomUtils.coordAtPos(geom, pos);
    }
}

export default AStarTraceNode;

File: comms/server/networks/a_star_trace_node.py
# Copyright: IQGeo Limited 2010-2023

from myworldapp.core.server.networks.myw_trace_node import MywTraceNode
from myworldapp.core.server.base.geom.myw_geo_utils import geodeticDistanceBetween


class AStarTraceNode(MywTraceNode):
    """
    Trace node for use with A* algorithm

    Adds property min_possible_dist which stores the minimum possible total length
    of the path to a stop node (if doing shortest path). This is computed assuming nodes
    a located in euclidean space (see web for description of A*)"""

    def __lt__(self, other):
        """
        Comparison operator (used in heap operations)
        """
        # Subclassed to support ordering by distance to stop nodes (part of the A* algorithm implementation)

        if hasattr(self, "min_possible_dist"):
            return self.min_possible_dist < other.min_possible_dist

        return self.dist < other.dist

    def minDistanceTo(self, geoms):
        """
        Distance from self to nearest vertex on GEOMS, in m (0.0 if not known)

        GEOMS is a set of shapely geometries
        """

        self_coord = self._stop_coord()

        if not self_coord:
            return 0.0

        min_dist = None
        for geom in geoms:
            for coord in geom.coords:
                dist = geodeticDistanceBetween(self_coord, coord)

                if min_dist == None or min_dist > dist:
                    min_dist = dist

        return min_dist

    def _stop_coord(self):
        """
        Position on self's feature at which self ends
        """

        if not self.parent:  # ENH: Do better if root node
            return None

        geom = self.featureGeom()
        if geom.geom_type == "Point":
            return geom.coords[0]

        # Find position of stop point along feature (as proportion of total length)
        # Remember: dist may have been computed from a stored length value
        if self.partial:
            pos = (self.dist - self.parent.dist) / (self.full_dist - self.parent.dist)
        else:
            pos = 1.0

        if not self.forward(geom):
            pos = 1.0 - pos

        return geom.geoCoordAtPos(pos)
